apps/deploy-web/src/context/WalletProvider/WalletProvider.tsx

```diff
@@ -175,40 +182,46 @@ export const WalletProvider = ({ children }) => {
   }
 
   async function signAndBroadcastTx(msgs: EncodeObject[]): Promise<boolean> {
-    setIsWaitingForApproval(true);
     let pendingSnackbarKey: SnackbarKey | null = null;
-    const enqueueTxSnackbar = () => {
-      pendingSnackbarKey = enqueueSnackbar(<Snackbar title="Broadcasting transaction..." subTitle="Please wait a few seconds" showLoading />, {
-        variant: "info",
-        autoHideDuration: null
-      });
-    };
     let txResult: TxOutput;
 
     try {
       if (user && managedWallet) {
-        enqueueTxSnackbar();
+        const mainMessage = msgs.find(msg => msg.typeUrl in MESSAGE_STATES);
+
+        if (mainMessage) {
+          setLoadingState(MESSAGE_STATES[mainMessage.typeUrl]);
+        }
+
         txResult = await txHttpService.signAndBroadcastTx({ userId: user.id, messages: msgs });
       } else {
+        const enqueueTxSnackbar = () => {
+          pendingSnackbarKey = enqueueSnackbar(<Snackbar title="Broadcasting transaction..." subTitle="Please wait a few seconds" showLoading />, {
+            variant: "info",
+            autoHideDuration: null
+          });
+        };
+        setLoadingState("waitingForApproval");
         const estimatedFees = await userWallet.estimateFee(msgs);
         const txRaw = await userWallet.sign(msgs, {
           ...estimatedFees,
           granter: feeGranter
         });
 
-        setIsWaitingForApproval(false);
-        setIsBroadcastingTx(true);
+        setLoadingState("broadcasting");
         enqueueTxSnackbar();
         txResult = await userWallet.broadcast(txRaw);
 
-        setIsBroadcastingTx(false);
+        setLoadingState(undefined);
       }
 
       if (txResult.code !== 0) {
         throw new Error(txResult.rawLog);
       }
 
-      showTransactionSnackbar("Transaction success!", "", txResult.transactionHash, "success");
+      if (!managedWallet?.isWalletConnected) {
```

Comment:
This feels like a weird condition. If managedWallet is truthy, there will never be a wallet connected. Looking at the hook, we set it to if the managed wallet is thruthy in a ternary for the wallet value. Perhaps we could just check if the userWallet is defined or has an address.